### PR TITLE
KIALI-478 aggregated health

### PR DESCRIPTION
**Pull request by jotak**
Some refactoring needed here to simplify coding:
- Introducing "business" layer between handlers and model. This layer is responsible for client API connections (prom, k8s) as well as some part in model building when it was tricky to do it directly from the model itself.
- New mocks for k8s and prom clients; injectable in business layer for tests

For Health:
- Changed model to accept several sources: envoy and each pod deployment status
- Set health directly in Service model in order to prevent duplicate round-trips from UI to fetch deployment status (another options would have been cache, but... not for the moment I think)

### CODE REVIEW

File: business/services_test.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	assert.Equal(int32(2), reviewsOverview.Health.DeploymentStatuses[0].AvailableReplicas)
+	assert.Equal(int32(3), reviewsOverview.Health.DeploymentStatuses[0].Replicas)
+	assert.Equal(int32(1), reviewsOverview.Health.DeploymentStatuses[1].AvailableReplicas)
+	assert.Equal(int32(2), reviewsOverview.Health.DeploymentStatuses[1].Replicas)
+	assert.Equal(model.SampleValue(6.5), reviewsOverview.RequestCount)
+	assert.Equal(model.SampleValue(1.5), reviewsOverview.RequestErrorCount)
+	assert.Equal(model.SampleValue(1.5/6.5), reviewsOverview.ErrorRate)
+
+	assert.Equal("httpbin", httpbinOverview.Name)
+	fmt.Printf("%v\n", httpbinOverview)
```
Comment (jotak) [inline business/services_test.go]: to remove

**Comment by lucasponce** [comment]
After a first review, I have some preliminar comments.

I see and I agree that increasing complexity in the handlers package is not good, so extract that is a good movement.

But, I am not sure if I agree with having two separate "models" and "business" package, I understand the split, but for me model + logic should be in the same package.

Perhaps the naming is confusing here and one of your first proposal "services" would help, or at least for me it has a better meaningful name.

I think "models" package purpose is right, as we discussed to have this "kiali" layer to have its own model, instead to propagate directly the "prometheus" and the "kubernetes" model.

But, as same way we don't split "prometheus" or "kubernetes" model and services in different packages, I guess we can continue with that strategy for consistency.

This doesn't mean that we should put everything in the same file, we can use some strategy in that way, to separate models and business in different files, but under the same package (indeed, we started doing that in the kubernetes package).

So, after read the code locally and think, having the "business" and "models" package separated confuses to me, instead to show more clarity.

I think one single package for the same purpose (i.e. "services" / "kiali_services" or similar) would work.

**Comment by lucasponce** [comment]
@jotak my rationale for this is that package should offer a model and an API to abstract (at least, the attempt) the complexity to the layer that uses that API.

Just to summarize:
- "prometheus" package was created to place there the complexity of the model of p8s, and doing the computation there, and return a calculated value. So, from kiali, I should use these layer to not have to deal with the details of the p8s model.
- "kubernetes" was created for same rationale, to query to k8s, and istio, and abstract the underlying details.

When we had this, we discussed what model we want to propagate over the handlers, and the discussion was to use an own model "models" to use this in the kiali API.

For me, it makes sense to follow a similar criteria, unfortunately, graph didn't follow it, so we have two different criteria here, and the whole architecture grows without a common design.

Personally, I think setting these abstractions are good, I like that the handlers have to invoke a simple API in the business layer, just passing parameters, and not doing the calculation in the handlers code, which I agree.

And yes, this refactoring is mainly how to rename, and more than naming, is to follow same criteria, if not at the end code will work, but looking at the code, people won't understand the design criteria.

So, as commented, my point, how I understand is more clear is to setting that abstraction layer in a single package, I think. Whatever name is, no matter which name we put, probably other will choose a different one. But conceptually, handlers should invoke a single package, I guess not splitted in two.

At least, is how I would do it.

But, again, this code will work, another proposal will work as well, and which approach is better/worse is just a matter of taste. My point is to be consistent with the previous criteria used, but it was broken since the beginning when we started.

**Comment by jotak** [comment]
@lucasponce ok the we agree on the essential. Just a remark on having `kubernetes` and `prometheus` packages as abstractions to build the model: the pitfall would be when model needs to be built from several sources, like it happens now with Health. There must be a place to fetch both prom & kube.

File: business/health.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package business
+
+import (
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+	"github.com/kiali/kiali/prometheus"
+	"k8s.io/api/apps/v1beta1"
```
Comment (jmazzitelli) [inline business/health.go]: Trivial comment - all our kiali source code orders the imports with:

```
import (
  // std imports first (like "net/http", etc)
  // blank line
  // third-part imports (like "k8s.io" or "glog")
  // blank line
  // our kiali imports (like our "config")
```

The blank lines allow gofmt to organize each section. We just have to put them in that order so its easy for a person reading the code to find imports quickly with their eyes.

This is how all the kiali go code organizes the imports.

Comment (jotak) [reply business/health.go]: Weird, I thought gofmt would take care of it for me, but it doesn't. Maybe I missed to configure something?

**Comment by jotak** [comment]
Here is a change with package `business` merged into `models`. 

I'm still not convinced, it creates a big 20-files package with wider responsibilities where we could have stricter boundaries and responsibilities with almost no effort.

But I prefer to move on so if there's an agreement then let's go for it.

**Comment by xeviknal** [comment]
I agree with Joel and I wouldn't mix concepts under the same package. I would like to follow a One Single Responsibility approach and keep things stupidly simple. I think the more separated the harder to mix concepts.

In general, I agree with lines followed into this refactor.

**Comment by xeviknal** [comment]
I just did this analysis of responsibilities. I thought it would be useful somehow.
Are we agree on that?

Handler action:
x treat inputs
x call Service/Manager/Controller
x treat output

Service/Manager/Controller
x get info from different clients.
x call models/serializers/parsers for structuring data according what the current request asks for.

API client/services
x call specific API
x return raw data.

Models/Serializers/Parsers
x given raw data transform into kiali preferred format.

**Comment by lucasponce** [comment]
@xeviknal your criteria is right, but I guess that the current one is right, too :-).

Just an example: in "kubernetes" package, basically we offer a client, a set of types, and the services to invoke them. We have logic inside to abstract the details of calling in low level the k8s and istio API, and perform some filtering and aggregation, and provide to the user of this package the final result.

So, following your criteria, we should also splitted in two, to separate model and service logic.

Both criteria works and we are agree on the responsabilities described with the exception that I still think is good idea to put together the types + logic used.

Anyway, to unblock @jotak and to find a compromise I suggest to move "models" and "business" as subpackages of a single one.

With this approach, we can solve both concerns: to separate them for clarity, but maintain that both subpackages belongs to same abstraction layer, and the subpackages can be seen more as an implementation detail that can change in the future, but not the high abstraction layer.

Sorry for the extra work @jotak, but I think this discussion was healthy.

What do you think ?

**Comment by jotak** [comment]
Rebased & squashed, integrating @xeviknal patch about matching services and deployments

File: kubernetes/kubetest/mock.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package kubetest
+
+import (
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/stretchr/testify/mock"
```
Comment (lucasponce) [inline kubernetes/kubetest/mock.go]: If I am not wrong, I guess latest import should be the kiali ones.

File: prometheus/prometheustest/mock.go [hunk 4]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"time"
 
+	"github.com/kiali/kiali/prometheus"
```
Comment (lucasponce) [inline prometheus/prometheustest/mock.go]: Should be this import at the end of the import list ?

File: services/business/services.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
+package business
+
+import (
+	"strings"
+
+	"github.com/kiali/kiali/config"
```
Comment (lucasponce) [inline services/business/services.go]: Same here, I think kiali imports should be at the end.

File: services/business/services_test.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"k8s.io/api/apps/v1beta1"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kiali/kiali/config"
```
Comment (lucasponce) [inline services/business/services_test.go]: kiali imports should be at the end.

File: services/models/service.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
+package models
+
+import (
+	"github.com/kiali/kiali/kubernetes"
```
Comment (lucasponce) [inline services/models/service.go]: Here too.

**Comment by jotak** [comment]
@jmazzitelli I'm merging it as your comment were addressed previously

**Pull request merged**